telegram/ext/picklepersistence.py

```diff
@@ -142,63 +188,64 @@ def dump_file(filename: str, data: object) -> None:
         with open(filename, "wb") as file:
             pickle.dump(data, file)
 
-    def get_user_data(self) -> DefaultDict[int, Dict[object, object]]:
+    def get_user_data(self) -> DefaultDict[int, UD]:
         """Returns the user_data from the pickle file if it exists or an empty :obj:`defaultdict`.
 
         Returns:
-            :obj:`defaultdict`: The restored user data.
+            DefaultDict[:obj:`int`, :class:`telegram.ext.utils.types.UD`]: The restored user data.
         """
         if self.user_data:
             pass
         elif not self.single_file:
             filename = f"{self.filename}_user_data"
             data = self.load_file(filename)
             if not data:
-                data = defaultdict(dict)
+                data = defaultdict(self.context_types.user_data)
             else:
-                data = defaultdict(dict, data)
+                data = defaultdict(self.context_types.user_data, data)
             self.user_data = data
         else:
             self.load_singlefile()
-        return deepcopy(self.user_data)  # type: ignore[arg-type]
```

Comment:
good point … I don't know why exactly I changed this, but the deepcopy-logic was introduced in #1297 to make sure that in the `update_*_data` the checks like `if self.chat_data[uid] == data` don't just compare the same object to itself (thus falsely skipping the updating).
By now this is already covered by the `replace/insert_bot` logic, so doing it twice is indeed overkill. Then again, I do have the idea to make the `replace/insert_bot` logic optional in which case you'd need it again. In any case #1297 didn't add/adjust any tests. So I'll remove the deepcopy-logic within PicklePersistence for now to avoid calling it twice but add tests that make sure that things work as expected.

Reply:
Then I guess it's fine to remove it, the tests covering this should be enough for now.

Reply:
Done, also for `DictPersistence`. Tests are now build in such a way that they fail if `replace/insert_bot` is skipped. Also added a note in `BasePersistence` about that